src/puffin/src/file_format/writer/file.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::{io, mem};
+
+use async_trait::async_trait;
+use futures::{AsyncRead, AsyncWrite, AsyncWriteExt};
+use snafu::ResultExt;
+
+use crate::blob_metadata::{BlobMetadata, BlobMetadataBuilder};
+use crate::error::{Result, WriteSnafu};
+use crate::file_format::writer::footer::FooterWriter;
+use crate::file_format::writer::{Blob, PuffinAsyncWriter, PuffinSyncWriter};
+use crate::file_format::MAGIC;
+
+/// Puffin file writer, implements both [`PuffinSyncWriter`] and [`PuffinAsyncWriter`]
+pub struct PuffinFileWriter<W> {
+    /// The writer to write to
+    writer: W,
+
+    /// The properties of the file
+    properties: HashMap<String, String>,
+
+    /// The metadata of the blobs
+    blob_metadata: Vec<BlobMetadata>,
+
+    /// The offset of the next blob
+    next_blob_offset: u64,
+}
+
+impl<W> PuffinFileWriter<W> {
+    pub fn new(writer: W) -> Self {
+        Self {
+            writer,
+            properties: HashMap::new(),
+            blob_metadata: Vec::new(),
+            next_blob_offset: 0,
+        }
+    }
+
+    fn create_blob_metadata(
+        &self,
+        typ: String,
+        properties: HashMap<String, String>,
+        size: u64,
+    ) -> BlobMetadata {
+        BlobMetadataBuilder::default()
+            .blob_type(typ)
+            .properties(properties)
+            .offset(self.next_blob_offset as _)
+            .length(size as _)
+            .build()
+            .expect("Required fields are not set")
+    }
+}
+
+impl<W: io::Write> PuffinSyncWriter for PuffinFileWriter<W> {
+    fn set_properties(&mut self, properties: HashMap<String, String>) {
+        self.properties = properties;
+    }
+
+    fn add_blob<R: io::Read>(&mut self, mut blob: Blob<R>) -> Result<()> {
+        self.write_header_if_needed_sync()?;
+
+        // write the blob into the writer
+        let size = io::copy(&mut blob.data, &mut self.writer).context(WriteSnafu)?;
+
+        let blob_metadata = self.create_blob_metadata(blob.blob_type, blob.properties, size);
+        self.blob_metadata.push(blob_metadata);
+
+        self.next_blob_offset += size;
+        Ok(())
+    }
+
+    fn finish(&mut self) -> Result<()> {
```

Comment:
Non-blocking:  is it necessary to call the `finish` when `drop` the struct automatically?

Reply:
it's really hard to do async drop in Rust